### PR TITLE
refactor(store): variadic tuple types for on

### PR DESCRIPTION
**Pull request by timdeschryver**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Closes #

## Does this PR introduce a breaking change?

Not that I know of 

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by timdeschryver** [comment]
Oh right, first #2794 has to be merged in 😅

File: modules/store/src/reducer_creator.ts [hunk 1]
```diff
@@ -11,177 +11,25 @@ export interface OnReducer<S, C extends ActionCreator[]> {
...
-  ...args: (ActionCreator | Function)[]
-): { reducer: Function; types: string[] } {
-  const reducer = args.pop() as Function;
+export function on<
+  C extends ActionCreator[],
+  S,
+  OR extends OnReducer<S, C> = OnReducer<S, C>,
+  AR extends ActionReducer<S> = ActionReducer<S>
+>(...args: [...C, OR]): On<S> {
+  const reducer = (args.pop() as Function) as AR;
```
Comment (markostanimirovic) [inline modules/store/src/reducer_creator.ts:32]:
```suggestion
export function on<
  Creators extends ActionCreator[],
  State,
  Reducer extends OnReducer<State, Creators>
>(...args: [...creators: Creators, reducer: Reducer]): On<State> {
  const reducer = (args.pop() as Function) as ActionReducer<State>;
```

Comment (markostanimirovic) [reply modules/store/src/reducer_creator.ts:32]: @timdeschryver
Btw, Prettier should be updated to [v2.1.0](https://prettier.io/blog/2020/08/24/2.1.0.html#support-for-typescript-40) or higher to support new TypeScript 4 features 🙂

Comment (timdeschryver) [reply modules/store/src/reducer_creator.ts]: Good suggestion, thanks!
I'll update prettier in #2794

**Review by alex-okrushko — APPROVED** [review]
Nice!

**Review by brandonroberts — APPROVED** [review]

**Comment by bbaia** [comment]
👍 
Can we do the same for `createSelector`?

**Comment by timdeschryver** [comment]
@bbaia Yes, we're looking into it, but we're hitting against some restrictions of TS.
It seems like TS will resolve this in one of the upcoming versions, once that's done we can revisit `createSelector`.

**Pull request merged**